### PR TITLE
Implement Reputation Log (#180)

### LINKED ISSUE #180: Reputation Listing
Make it possible for a user to see how many Reps the user has received and given.

### PR DESCRIPTION
**Pull request by AbnerSquared**
## Related Issue
This pull request implements the missing methods required to create reputation logs and listings.

Fixes #180 

## Changes
- Created Paginator, which handles creating pages for collections.
- Modified EmbedConstructor, IDiscordMessages, ProfileService, ProfileCommand, DiscordMessagesAdapter, IUserReputationProvider, UserReputationProvider, IMiunieUserProvider, MiunieUserProvider, PersistentStorage, and IPersistentStorage to include methods relating to getting all users, and getting reputation.
- Created ReputationEntry and ReputationType to help sort reputation.
- Modified Strings.resx and PhraseKey to include two new phrases: `USER_EMBED_REP_LOG_TITLE` and `USER_EMBED_REP_LOG_EMPTY`

## Expected Feedback
- Determine if created methods are in their intended locations.

### CODE REVIEW

**Comment by AbnerSquared** [comment]
If needed, I can go through one more polish check and remove redundant comments and fix some line spacing.

File: src/Miunie.Core/Entities/ReputationEntry.cs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+            Type = type;
+            FromInvoker = fromInvoker;
+        }
+
+        public ulong TargetId { get; set; }
+        public string TargetName { get; set; }
+        public DateTime GivenAt { get; set; }
+        public ReputationType Type { get; set; }
+
+        public bool FromInvoker { get; set; }
```
Comment (DraxCodes) [inline src/Miunie.Core/Entities/ReputationEntry.cs]: I feel this reads better as;

```cs
public bool IsFromInvoker { get; set; }
```

File: src/Miunie.Core/Providers/MiunieUserProvider.cs [hunk 2]
```diff
@@ -1,7 +1,9 @@
 using Miunie.Core.Storage;
+using System.Collections.Generic;
 
 namespace Miunie.Core.Providers
 {
+    // TODO: Provide a way to read all users, or store repGiven as a property in MinuieUser.
```
Comment (DraxCodes) [inline src/Miunie.Core/Providers/MiunieUserProvider.cs]: Remove Comment, :smile:

File: src/Miunie.Core/Providers/MiunieUserProvider.cs [hunk 3]
```diff
@@ -31,7 +33,12 @@ public void StoreUser(MiunieUser user)
             }
         }
 
-        private MiunieUser EnsureExistence(
+        public IEnumerable<MiunieUser> GetAllUsers()
+        {
+            return _persistentStorage.RestoreAll<MiunieUser>();
+        }
+
+        private MiunieUser Exists(
```
Comment (DraxCodes) [inline src/Miunie.Core/Providers/MiunieUserProvider.cs]: Not sure if an accidental change but makes more sense to have it as EnsureExistence or at-least EnsureExists

File: src/Miunie.Discord/Embeds/EmbedConstructor.cs [hunk 4]
```diff
@@ -19,6 +52,7 @@ public static Embed ToEmbed(this MiunieUser mUser, ILanguageProvider lang)
                 .AddField(lang.GetPhrase(PhraseKey.USER_EMBED_NAME_TITLE.ToString()), mUser.Name)
                 .AddField(lang.GetPhrase(PhraseKey.USER_EMBED_REALNESS_TITLE.ToString()), realnessPhrase, true)
                 .AddField(lang.GetPhrase(PhraseKey.USER_EMBED_REP_TITLE.ToString()), mUser.Reputation.Value.ToString(), true)
+                //.AddField(lang.GetPhrase(PhraseKey.USER_EMBED_REP_GIVEN_TITLE.ToString()), "", true) // This needs to refer to all users and calculate the best info
```
Comment (DraxCodes) [inline src/Miunie.Discord/Embeds/EmbedConstructor.cs]: Remove comment :smile:

File: src/Miunie.Discord/Embeds/Paginator.cs [hunk 5]
```diff
@@ -0,0 +1,82 @@
+﻿using Discord;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace Miunie.Discord.Embeds
+{
+    // This handles selecting portions of a list to display as a page.
```
Comment (DraxCodes) [inline src/Miunie.Discord/Embeds/Paginator.cs]: Remove Comment :smile:

File: src/Miunie.Discord/Embeds/Paginator.cs [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+            return (int)Math.Ceiling((double)collectionSize / pageSize);
+        }
+
+        public static EmbedBuilder PaginateEmbed<T>(IEnumerable<T> set, EmbedBuilder embed, int index, int pageSize)
+        {
+            return embed.WithDescription(Paginate(set, index, pageSize))
+                .WithFooter($"{(string.IsNullOrWhiteSpace(embed.Footer?.Text) ? "" : $"{embed.Footer?.Text} | ")}{GetPageFooter(index, set.Count(), pageSize)}");
+        }
+
+        // TODO: Correlate with ILanguageProvider
```
Comment (DraxCodes) [inline src/Miunie.Discord/Embeds/Paginator.cs]: Remove Comment :smile:

**Comment by petrspelos** [comment]
Thank you, very much, @AbnerSquared! ❤️ 

We'll take it from here. 😊

bitmoji

**Comment by petrspelos** [comment]
Turns out there are some major changes required to the reputation system.
Thank you, @AbnerSquared for the PR, we'll merge it for the time being. ❤️ 

bitmoji

**Comment by AbnerSquared** [comment]
Alright. Let me know what needs to be adjusted, and I'm more than willing to fix them as needed!

**Pull request merged**